File: Chapter2/Go/linear_search.go
package main

import "fmt"

type result struct {
	found   bool
	index   int
	element int
}

func linearSearch(array []int, x int) result {
	var found bool
	for idx, elem := range array {
		if elem == x && !found {
			found = true
			return result{
				found,
				idx,
				elem,
			}
		}
	}
	return result{}
}

func main() {
	array := []int{1, 2, 3, 4, 5, 6}
	result := linearSearch(array, 6)
	fmt.Println(result)
	badResult := linearSearch(array, 1000)
	fmt.Println(badResult)
}

File: Chapter2/Python/linear_search.py


from typing import Any


def linear_search(A: list[Any], x: int) -> tuple: 
    found = False 
    for idx, elem in enumerate(A):
        if elem == x and not found:
            found = True
            return (found, idx, elem)
    return (found, None, None)

if __name__ == "__main__":
    array = [1,2,3,4,5,6]
    result = linear_search(array, 4)
    print(result)
    bad_result = linear_search(array, 100)
    print(bad_result)
